### PR TITLE
Tasks 1-17 solved

### PR DESCRIPTION
**Pull request by IbrayevRamil**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/.idea/
+/build/
+/.gradle/
```
Comment (nkarulin) [inline .gitignore]: .gitignore should not be in source repository

File: src/main/java/com/epam/javacourses2016/task1/SolverTask1.java [hunk 2]
```diff
@@ -16,7 +20,32 @@
...
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        List<String> reversedList = new ArrayList<>(list);
+        Collections.reverse(reversedList);
+        try {
+            if (!output.exists()) {
+                output.createNewFile();
+            }
+            FileWriter fw = new FileWriter(output.getAbsoluteFile());
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task1/SolverTask1.java]: Why don't you use try-with-resource on Writer? It should also be closed after use

File: src/main/java/com/epam/javacourses2016/task10/SolverTask10.java [hunk 3]
```diff
@@ -17,6 +21,21 @@
...
+        try {
+            lines = Files.readAllLines(input.toPath(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        HashMap<String, Integer> words = new HashMap<>();
+        for(String line: lines){
+            if (words.containsKey(line)) {
+                words.put(line, words.get(line) + 1);
+            }
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task10/SolverTask10.java:43]: better keep `else` on the same line
```java
if (...) {
   //...
} else {
   //...
}

File: src/main/java/com/epam/javacourses2016/task11/SolverTask11.java [hunk 4]
```diff
@@ -16,16 +16,26 @@
      * @param peoples Список с именами участников.
      * @return Имя последнего оставшегося.
      */
-    public String emulate(ArrayList<String> peoples) {
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task11/SolverTask11.java:19]: Don't you feel confused about copy-paste in those methods?)

Realization for ArrayList is not effective

File: src/main/java/com/epam/javacourses2016/task12/SolverTask12.java [hunk 5]
```diff
@@ -16,7 +19,30 @@
...
+            else if (integers.get(j) >= value && j <= count-1) {
+                for (int i = count; i < integers.size(); i++) {
+                    if (integers.get(i) < value) {
+                        Collections.swap(integers, j, i);
+                        break;
+                    }
+                }
+            }
+        }
+        return integers;
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task12/SolverTask12.java:46]: Can be simplified by using custom Comparator in sorted collection

File: src/main/java/com/epam/javacourses2016/task17/SolverTask17.java [hunk 6]
```diff
@@ -19,7 +19,51 @@
...
+        //List<Point2D> crossedSegments = new ArrayList<>();
+        TreeMap<Integer[], Point2D> crossedSeg = new TreeMap<>();
+        Segment[] segArray = segments.toArray(new Segment[segments.size()]);
+        for (int i = 0; i < segArray.length; i++) {
+            for (int j = i; j < segArray.length; j++) {
+                Point2D crossDot = crossDot(segArray[i], segArray[j]);
+                if (crossDot != null) {
+                    crossedSeg.put(new Integer[]{i+1,j+1}, crossDot);
+                }
+            }
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task17/SolverTask17.java:32]: nice but slow:snail: O(n^2)

File: src/main/java/com/epam/javacourses2016/task3/SolverTask3.java [hunk 7]
```diff
@@ -15,8 +16,23 @@
...
+        }
+        for (int i = lines.size() - 1; i >= 0; i--) {
+            for (int j = 0; j < i; j++) {
+                if (lines.get(j).length() > lines.get(j + 1).length()) {
+                    String t = lines.get(j);
+                    lines.set(j, lines.get(j+1));
+                    lines.set(j+1, t);
+                }
+            }
+        }
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task3/SolverTask3.java:34]: Don't use bubble sort in practice:)
Collections.sort is much better

File: src/main/java/com/epam/javacourses2016/task5/SolverTask5.java [hunk 8]
```diff
@@ -18,7 +21,15 @@
...
-        return 0;
+        BigDecimal Syx = new BigDecimal(0);
+        for (Measurement measurement : measurements) {
+            Syx = Syx.add(new BigDecimal(measurement.getCurrent()*measurement.getVoltage()));
+        }
+        BigDecimal Sxx = new BigDecimal(0);
+        for (Measurement measurement : measurements) {
+            Sxx = Sxx.add(new BigDecimal(measurement.getCurrent()*measurement.getCurrent()));
+        }
+        BigDecimal R = Syx.divide(Sxx,6);
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task5/SolverTask5.java]: What the magic number?

File: src/main/java/com/epam/javacourses2016/task5/SolverTask5.java [hunk 9]
```diff
@@ -18,7 +21,15 @@
...
+        return R.doubleValue();
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task5/SolverTask5.java]: Are you sure it is method of least squares?

File: src/main/java/com/epam/javacourses2016/task8/SolverTask8.java [hunk 10]
```diff
@@ -16,6 +18,30 @@
...
+            char current = string.charAt(i);
+            if (current == '{' || current == '(' || current == '[') {
+                stack.push(current);
+            }
+            if (current == '}' || current == ')' || current == ']') {
+                if (stack.isEmpty()) {
+                    return false;
+                }
+                char last = stack.peek();
+                if (current == '}' && last == '{' || current == ')' && last == '(' || current == ']' && last == '[') {
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task8/SolverTask8.java:37]: can be refactored with enums
Keep data away from code

File: src/main/java/com/epam/javacourses2016/task9/SolverTask9.java [hunk 11]
```diff
@@ -16,7 +20,29 @@
      * @param input Анализируемый файл.
      * @return Множество полученных слов.
      */
+
     public HashSet<String> getUniqueWords(File input) {
-        return null;
+        List<String> lines = null;
+        try {
+            lines = Files.readAllLines(input.toPath(), StandardCharsets.UTF_8);
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task9/SolverTask9.java]: :snail: 
We dont need that buffer actually, single Set with custom comparator should be enough

File: src/test/java/com/epam/javacourses2016/task1/Task1Test.java [hunk 12]
```diff
@@ -12,7 +12,7 @@
 
 public class Task1Test {
 
-    @Test(enabled = false, dataProvider = "lines")
+    @Test(enabled = true, dataProvider = "lines")
```
Comment (nkarulin) [inline src/test/java/com/epam/javacourses2016/task1/Task1Test.java:15]: Please enable other tests too

**Comment by IbrayevRamil** [comment]
task 15, 16 need some fix

File: src/main/java/com/epam/javacourses2016/task13/Graph.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        nodes = new HashSet<>();
+        for (int i = 0; i < NUMBER_NODES; i++) {
+            Node node = new Node(i);
+            nodes.add(node);
+        }
+    }
+
+    @Override
+    public void addEdge(int first, int second) {
+        for (Node node : nodes) {
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task13/Graph.java:26]: Set or Map give us O(logn) or even O(1) (for HashMap or HashSet with no collisions) on searching for element, but your implementations of Graph and Node don't use that speedup using iteration through collection with just O(n) :snail:

File: src/main/java/com/epam/javacourses2016/task13/Node.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
+package com.epam.javacourses2016.task13;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by Рамиль on 29.11.2016.
+ */
+public class Node {
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task13/Node.java:9]: equals/hashCode convention will help you to speed up your graph :rabbit2:

File: src/main/java/com/epam/javacourses2016/task14/MyNumberCollection.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Created by Рамиль on 30.11.2016.
+ */
+public class MyNumberCollection<T extends Number> implements AbstractCollectionCreator.NumberCollection<T> {
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task14/MyNumberCollection.java:11]: extending java `AbstractCollection` will help you to save some lines of code

File: src/main/java/com/epam/javacourses2016/task15/IFile.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+            while (scanner.hasNext()) {
+                if (scanner.hasNextDouble()) {
+                    lineDots.add(scanner.nextDouble());
+                }
+                final int MAX_LINE_COORDINATES = 4;
+                if (lineDots.size() == MAX_LINE_COORDINATES) {
+                    final int A_X_LINE_COORDINATE = 0;
+                    final int A_Y_LINE_COORDINATE = 1;
+                    final int B_X_LINE_COORDINATE = 2;
+                    final int B_Y_LINE_COORDINATE = 3;
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task15/IFile.java]: These guys are constants. They should be static final class fields.

File: src/main/java/com/epam/javacourses2016/task16/Cell.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Created by Рамиль on 30.11.2016.
+ */
+public class Cell {
+
+    private final Point2D CENTER;
+    private final Point2D TOP_LEFT_CORNER;
+    private final Point2D BOT_LEFT_CORNER;
+    private final Point2D TOP_RIGHT_CORNER;
+    private final Point2D BOT_RIGHT_CORNER;
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task16/Cell.java]: But these guys are not constants, they are just final fields and should not be written in CAPITAL.

File: src/main/java/com/epam/javacourses2016/task16/Cell.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    private final double CELL_SIZE = 1.0;
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task16/Cell.java]: This guy is constant it should be static

File: src/main/java/com/epam/javacourses2016/task16/IFile.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * Created by Рамиль on 30.11.2016.
+ */
+public class IFile implements SolverTask16.IFileWithPoints {
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task16/IFile.java]: _I_ prefix means _Interface_, actually it is not common for Java more C++ style. But anyway it should not be in naming of classes.

File: src/main/java/com/epam/javacourses2016/task16/IFile.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    @Override
+    public SortedMap<Point2D, Double> getPoints() {
+        SortedMap<Point2D, Double> cells = new TreeMap<>();
+        try (Scanner scanner = new Scanner(new FileReader(file))) {
+            ArrayList<Double> coords = new ArrayList<>();
+            while (scanner.hasNext()) {
+                if (scanner.hasNextDouble()) {
+                    coords.add(scanner.nextDouble());
+                }
+                final int DOT_COORDINATES_AND_DIST = 3;
```
Comment (nkarulin) [inline src/main/java/com/epam/javacourses2016/task16/IFile.java]: Constant, put it in class header as static

**Comment by nkarulin** [comment]
Nice! But now we need to wait for tests covered all your tasks.
Better keep PR only with covered tasks to merge it earlier.

Also, resolve conflict please

**Comment by IbrayevRamil** [comment]
Finally did it 😁

**Comment by nkarulin** [comment]
Just noticed. Please change target branch from master to yours.

**Comment by nkarulin** [comment]
master was updated

**Pull request merged**